---
title: "Rarefied ordination v2"
author: "Harald Brolin"
date: "September 25, 2018"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, echo = FALSE)
```


```{r library, echo=TRUE, message=FALSE, warning=FALSE}
library("phyloseq"); packageVersion("phyloseq")
library("ggplot2"); packageVersion("ggplot2")
library("tidyverse")
library("gridExtra")
library("knitr")
library("Hmisc")
library("metagenomeSeq")
library("EnvStats")
library("vegan")
```



```{r load}
df <- readRDS(file = "../phyloseq_v2.rds")
set.seed(2018)
df@sam_data$sequence_depth <- apply(df@otu_table, 2, function(x) (sum(x))) # Ad sequencing depth for each sample
```

#### Normalization

Inspiration on normalization can be found here: [link](https://www.researchgate.net/publication/314194282_Normalization_and_microbial_differential_abundance_strategies_depend_upon_data_characteristics)

##### Rarefying
In short rarefying more clearly clusters samples according to biological origin than other normalziation methods. Rarefying does not increase the FDR in itself but does reduce the sensitivity since a part of the data is removed. For "groups" (hard to tell if the meaning of groups refers to ex. sample sites or sequencing run) with a large difference (~10x) in average library size rarefying lowers the FDR. 

Conclusion from this paper: [link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5123278/)

"..The effect of library size normalization was by far the lowest, especially between no normalization, CSS, DESeq2, and TMM normalization.."
"When exploring beta diversity of microbiome data, analysts should carefully consider their choice of count transformation and distance metric, the latter having the largest impact on results...""

#### Important to note:
To evaluate the cluster separation from a beta-diversity analysis the adonis-function in the vegan-package can be used. Link to example article (same as above): [link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5123278/)

### Beta-diversity

The phyloseq tutorial recommends to remove low-occurence, porly represented OTUs. In our case we have ASVs, I do not know how this affects the filtering. It is important to think about the method of count-transformation, sample depth normalization and distance (diversity) measurment, it is also important to decide upon which filtering should be used. Singletons could be removed, note that and ASV singleton and an OTU singleton are quite different.

##### First beta analysis, no ASVs removed and rarefied, using Bray-Curtis, Unifrac-U, Unifrac-W.
```{r transform and create new df based on different distances, echo=TRUE, message=FALSE}
# Rarefy to Overby131205 the sample with the lowest sequence depth
# rarefy without replacement as per the R.Knight article
df <- rarefy_even_depth(df,
                        df@sam_data$sequence_depth %>% min, # get minimum sample sequence depth 
                        rngseed = 2018,
                        replace = FALSE,
                        trimOTUs = TRUE,
                        verbose = TRUE)

u_unifrac <- ordinate(df, "PCoA", "uunifrac") # Create an ordination 
```


```{r plot scatter uuniFrac}
p2 <-
  plot_ordination(df, u_unifrac, color = "location", shape = "season") + ggtitle("Unweighted-Unifrac,  rarefied data"); p2

ggsave(plot = p2, filename = "../../analysis_images/2_uufrac_rarefy.png")
```

#### Conclusion
Unweighted UniFrac has the highest level of separation for the phlogenetic distance measurments, but has the lowest percentage explained variation (sum = ~9%). 

Add barplot to the existing plot. **Note that I have manually aligned the bar-plots, this could cause a slight mis-alignment**.

```{r get barplots based on axis}
# Since the rownames are in the same order
locations <- df@sam_data$location %>% unique() %>% as.vector() # Get location names
seasons <- df@sam_data$season %>% unique() %>% as.vector() # Get season names

get_axis <- function(variable, axis, column) {
  # This function gets rows dependet on vairable (Stockholm etc. or Autumn etc.) and the corresponding Axis
  # The function returns a dataframe with samplenames, Axis-values and category (either season or location)
  rows <- which(df@sam_data[, column] == variable)
  cbind(
    (u_unifrac$vectors[rows, axis]),  
    rep(variable, row %>% length)) %>% 
  as.data.frame() %>% 
  setNames(., nm = c(axis, column))  #Lägg till mnumeric char
}

location_df <- # For each location get the corresponding Axis 2 values, bind into data frame
  lapply(locations, get_axis, axis = "Axis.2", column = "location") %>%
  do.call(rbind, .) # Cbind all dataframes from the lapply

location_df$Axis.2 <- location_df$Axis.2 %>% as.character() %>% as.numeric() # Covert from factor to numeric
location_df$location <- factor(location_df$location,
                                   levels = c("Borås", "Härnösand", "Motala", "Östersund", "Stockholm", "Trollhättan"))

season_df <- # For each season get the corresponding Axis 1 values, bind into data frame
  lapply(seasons, get_axis, axis = "Axis.1", column = "season") %>%
  do.call(rbind, .)

season_df$Axis.1 <- season_df$Axis.1 %>% as.character() %>% as.numeric() 
```





```{r multiplot}

blankPlot <-  # Create an empty plot
  ggplot()+geom_blank(aes(1,1)) +
  theme(plot.background = element_blank(), 
   panel.grid.major = element_blank(),
   panel.grid.minor = element_blank(), 
   panel.border = element_blank(),
   panel.background = element_blank(),
   axis.title.x = element_blank(),
   axis.title.y = element_blank(),
   axis.text.x = element_blank(), 
   axis.text.y = element_blank(),
   axis.ticks = element_blank()
     )

box_location <- # Create boxplot of location, color by location 
  ggplot(location_df, aes(y = Axis.2, x = location)) +
  geom_boxplot(aes(fill = location), show.legend = FALSE) +
  theme(plot.margin = unit(c(0.15,0,1.1,0.2), "cm"), # Sync the images c(top, right, bottom, left) 
        axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        legend.position="none",
        panel.background=element_blank(),
        panel.border=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        plot.background=element_blank()) 

box_season <- # Create boxpålot of seasons
  ggplot(season_df, aes(y = Axis.1, x = season)) +
  geom_boxplot() +
  coord_flip() +
  theme(plot.margin = unit(c(0,3.2,0,0.12), "cm"),# Sync the images c(top, right, bottom, left)
        axis.text.x = element_blank(),
        axis.ticks=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        legend.position="none",
        panel.background=element_blank(),
        panel.border=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        plot.background=element_blank())

p2 <- # Create the maion ordination plot
  plot_ordination(df, u_unifrac, color = "location", shape = "season") +
  theme(legend.title = element_blank(),
        legend.key = element_blank())

p3 <- grid.arrange(blankPlot, # Empty plot
             box_season, # Boxplot season
             box_location, # boxplot location
             p2, # Scatterplot
             ncol = 2,
             widths=c(1.2, 6),
             heights=c(1.2, 6),
             top = ("Unweighted-UniFrac")); p3

ggsave(plot = p3, filename = "../../analysis_images/2_barplot_uu_ordination_rarefy.png")
```


Next we want to investigate the batch effect after rarefying the data. Below we'll visualize clustering based on run date and sample sequence depth (this shows the pre-rarefaction depth).


```{r plot run date sequence depth}
p1 <- 
  plot_ordination(df, u_unifrac, color = "run_date", shape = "season") +
  ggtitle("Unweighted-Unifrac, rarefied data"); p1
p2 <- 
  plot_ordination(df, u_unifrac, color = "sequence_depth", shape = "season") +
  ggtitle("Unweighted-Unifrac, rarefied data"); p2 

ggsave(plot = p2, filename = "../../analysis_images/2_app_uufrac_rarefy_sequence_run.png")
ggsave(plot = p2, filename = "../../analysis_images/2_app_uufrac_rarefy_depth.png")
```

#### Conclusion
The clustering based on sequence run date does exist, one clear cluster is 2015 run dates which both are located to the left in the plot. Otherwise we see relatively random pattern spread over locations. Note, it is hard to differentiate between run_date and season since theese are cofounded. The clustering based on sequence depth seems to be almost non-existing (middle-top plot we can see a small clustering), this is the expected result from rarefying. 

From a previos mark-down object *ordination_plot.Rmd* we saw that bray-curtis was heavily dependet on sample sequence depth, but with rarefied data this seems not to be the case.

#### Principal components
Though the unweighted UniFrac distance measurment seemed to separate the samples in the best way (sample site and season) the percent explained variation was relativley low. Looking at both bray-curtis and weighted UniFrac we have a higher percentage (22% resp. 37.7%), it could be the case that other principal-components separates the samples in a better way.

## permAnova test for factors

```{r permanova, echo=TRUE}
# options(knitr.table.format = "html", knitr.kable.NA = '') # For the kable output

metadata <- as(sample_data(df), "data.frame")

uu_distance_df <- distance(df, method = "uunifrac")
wu_distance_df <- distance(df, method = "wunifrac")

adonis_uunifrac <- vegan::adonis(uu_distance_df ~ location + season + run_date + season*location,
                                 data = metadata,
                                 permutations = 999,
                                 parallel = 20)
kable(adonis_uunifrac$aov.tab,
      digits = 3,
      align = "c",
      caption = "Adonis test, u-uniFrac ~ location + season + run_date + season:location")

adonis_wunifrac <- vegan::adonis(wu_distance_df ~ location + season + run_date + season*location, 
                                 data = metadata,
                                 permutations = 999,
                                 parallel = 20)
kable(adonis_wunifrac$aov.tab,
      digits = 3,
      align = "c",
      caption = "Adonis test, wu-uniFrac ~ location + season + run_date + season:location")
```

<!-- ```{r} -->
<!-- source("pairwise_adonis.R") -->


<!-- adonis_location <- pairwise.adonis(x = uu_distance_df %>% as.matrix(), -->
<!--                               factors = c(df@sam_data$location,df@sam_data$season, df@sam_data$run_date), -->
<!--                               perm = 999) -->

<!-- summary.pwadonis(adonis_location) -->
<!-- ``` -->

